<launch>

    <!-- launch PID Manager node -->
    <node name="self_balance" pkg="bbot_control" type="pid_control.py" output="screen">
        <remap from="imu" to="/bbot/imu"/>
        <remap from="cmd_vel" to="/bbot/mobile_base_controller/cmd_vel"/>
        <remap from="b_setpoint" to="/balancing_pid/setpoint" />
        <remap from="b_state" to="/balancing_pid/state" />
        <remap from="b_control_effort" to="/balancing_pid/control_effort" />
        <remap from="v_setpoint" to="/velocity_pid/setpoint" />
        <remap from="v_state" to="/velocity_pid/state" />
        <remap from="teleop" to="/bbot/teleop"/>
    </node>

    <!-- launch balancing PID node -->
    <node name="balancing_pid" pkg="pid" type="controller" output="screen" >
      <param name="Kp" value="-10.0" /> 
      <param name="Ki" value="-200.0" />
      <param name="Kd" value="-0.8"/>
      <param name="upper_limit" value="0.45" />
      <param name="lower_limit" value="-0.45" />
      <param name="windup_limit" value="0.1" />
      <param name="cutoff_frequency" value="20" />
      <param name="max_loop_frequency" value="105.0" />
      <param name="min_loop_frequency" value="95.0" />
      <param name="setpoint_timeout" value="-1.0" />
      <remap from="setpoint" to="~setpoint" />
      <remap from="state" to="~state" />
      <remap from="control_effort" to="~control_effort" />
      <remap from="pid_debug" to="~pid_debug" />
      <remap from="pid_enable" to="~pid_enable" />
    </node>

    <!-- launch velocity control PID node -->
    <node name="velocity_pid" pkg="pid" type="controller" output="screen" >
      <param name="Kp" value="0.1"/> 
      <param name="Ki" value="1.0" />
      <param name="Kd" value="0.0"/>
      <param name="upper_limit" value="0.09" />
      <param name="lower_limit" value="0.073" />
      <param name="windup_limit" value="0.1" />
      <param name="cutoff_frequency" value="20" />
      <param name="max_loop_frequency" value="105.0" />
      <param name="min_loop_frequency" value="95.0" />
      <param name="setpoint_timeout" value="-1.0" />
      <remap from="setpoint" to="/velocity_pid/setpoint" />
      <remap from="state" to="/velocity_pid/state" />
      <remap from="control_effort" to="/velocity_pid/buffer" />
      <remap from="pid_debug" to="~pid_debug" />
      <remap from="pid_enable" to="~pid_enable" />
    </node>

    <!-- launch node to publish setpoint msgs for the velocity PID in a constant frequency -->
    <node name="freq_pub" pkg="bbot_control" type="freq_pub.py" output="screen">
        <remap from="pub_topic" to="/balancing_pid/setpoint"/>
        <remap from="sub_topic" to="/velocity_pid/buffer"/>
    </node>

    <!-- launch twist_mux node -->
    <arg name="cmd_vel_out" value="/bbot/teleop"/>
    <arg name="config_topics" value="$(find bbot_control)/config/twist_mux.yaml"/>
    <node pkg="twist_mux" type="twist_mux" name="twist_mux" output="screen">
      <remap from="cmd_vel_out" to="$(arg cmd_vel_out)"/>
      <rosparam file="$(arg config_topics)" command="load"/>
      <rosparam param="locks">[]</rosparam>
    </node>

    <!-- launch balancing PID plot  -->
    <!-- <node name="bal_plot" pkg="rqt_plot" type="rqt_plot"
    args="/balancing_pid/control_effort/data /balancing_pid/state/data /balancing_pid/setpoint/data" /> -->

    <!-- launch velocity PID plot  -->
    <!-- <node name="pos_plot" pkg="rqt_plot" type="rqt_plot"
    args="/balancing_pid/setpoint/data /velocity_pid/state/data /velocity_pid/setpoint/data" /> -->

    <!-- <node name="rqt_reconfigure" pkg="rqt_reconfigure" type="rqt_reconfigure" /> -->
    <!-- <node name="rqt_robot_monitor" pkg="rqt_robot_monitor" type="rqt_robot_monitor" /> -->
</launch>